src/Operations.jl

```diff
@@ -1065,6 +1079,45 @@ function _resolve(ctx::Context, pkgs::Vector{PackageSpec}, preserve::PreserveLev
         targeted_resolve(ctx, pkgs, preserve)
 end
 
+has_unregistered_source(project::Project, pkg::PackageSpec) = pkg.uuid in keys(project.deps) && pkg.name in values(project.source)
+
+function collect_unregistered!(ctx::Context, pkg::PackageSpec, new_git::Vector{UUID}, unreg::Vector{PackageSpec})
+    projectfile = projectfile_path(source_path(pkg); strict=true)
+    projectfile === nothing && return # TODO best to error out here
+    src_project = Types.read_project(projectfile)
+    src_manifest = Types.read_manifest(manifestfile_path(source_path(pkg)))
+    for (name, uuid) in src_project.deps
+        new_pkg = nothing
+        entry = get(src_manifest, uuid, nothing)
+        if entry !== nothing && !is_tracking_registry(entry)
+            if is_tracking_path(entry)
+                path = Types.relative_project_path(ctx, project_rel_path(source_path(pkg), entry.path))
+                new_pkg = PackageSpec(;name=name, uuid=uuid, path=path, repo=entry.repo, tree_hash=entry.tree_hash)
+                # TODO handle missing path
+            else # tracking repo
+                new_pkg = Packagespec(;name=name, uuid=uuid, repo=entry.repo, tree_hash=entry.tree_hash)
+                Types.handle_repo_add!(ctx, new_pkg) && push!(new_git, new_pkg.uuid)
+            end
+        elseif (source = get(src_project.source, name, nothing)) !== nothing
+            new_pkg = PackageSpec(;repo=GitRepo(;source=source))
+            Types.handle_repo_add!(ctx, new_pkg) && push!(new_git, new_pkg.uuid)
+        end
+        if new_pkg !== nothing
+            push!(unreg, new_pkg)
+            collect_unregistered!(ctx, new_pkg, new_git, unreg)
```

Comment:
I wonder why you have to recurse? In principle, the manifest of `pkg` is likely to contain all URLs for unregistered dependencies, isn't it? Otherwise, such manifest cannot be instantiated.

Reply:
What about when you do an add on a url and that package has dependencies on other unregistered packages?

Reply:
In this case, I guess you'd need to parse the manifest file(s) contained in the newly to-be-added package?  But what I was commenting was that, if this new package has a manifest file, it is likely that you don't need to parse the manifest files of _dependencies_ of this newly added package.  Or at least you can try to minimize the I/O by stop recursing if all direct and transitive dependencies are found.  It could be an unnecessary premature optimization but don't need something like this to prevent infinite recursion anyway?

Reply:
@tkf Perhaps you missed it, but this PR also introduces a `source` field (as discussed in #492). So there could be no manifest files involved.

Reply:
Thanks for the clarification. I was confused since I came to this PR via another PR #1088.